### DISCUSSION TITLE
Compiling `owlkettle` in Windows

### DISCUSSION THREAD

**Discussion by RickBarretto** [can-lehmann/owlkettle#53] [Show and tell]
Hello there, I noticed that `owkettle` doesn't have any build instructions for Windows, nevertheless I wanted to compile using `Msys2`'s ecosystem. 

A priori, I had some problems while trying to compile, specially because of the `ld`. You can see the returned errors [in this thread from Discord][discord-thread].

## How I solved it

Basically, I removed this two lines: [`owlkettle/cairo#80`][cairo-line] and [`owlkettle/gtk#80`][gtk-line], and passed manually these arguments via `config.nims` with:

```nim
--passL: "-lgtk-4 -lpangowin32-1.0 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -lgdk_pixbuf-2.0 -lcairo-gobject -lcairo -lgraphene-1.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lintl"
```

For some reason, since I can't use `gorge("pkg-config --libs cairo")` into a nims file to pass the `pkg-config`'s result to `--passL`, so I did it manually in my terminal and pasted the result to `--passL`.

Also, build using `Msys2` or `Mingw` terminal, it's not working to me while using VsCode's terminal.

## Updates

So, I tried again and again, and ✨ **finally** ✨ found the error here, as you can see [in this Thread][discord-thread-2]. The above solution works, but if you want to understand what is happening, please read this.

Basically the problem is the `gorge` function. `gorge` is returning a multiline string what is breaking the `ld`'s parameters...

So, if you write
```nim
{.passl: "-lgtk-4 -lpangowin32-1.0 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -lgdk_pixbuf-2.0 -lcairo-gobject -lcairo -lgraphene-1.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lintl" .}
```

or 

```nim
import std/strutils
{.passl: gorge("pkg-config --libs gtk4").strip() .}
```
into your main file, and delete the `config.nims` it'll work as well.

The proof I found that `gorge` is returning a multiline was:

```nim
static:
  echo "\n." 
  echo gorge "pkg-config --libs gtk4"
  echo "."
```

That returns:

```
.
-lgtk-4 -lpangowin32-1.0 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -lgdk_pixbuf-2.0 -lcairo-gobject -lcairo -lgraphene-1.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lintl

.
```

---

## Side note: How I installed gtk4

Basically, as I said, I'm using `Msys2`, so I just followed this the [official installation instructions][gtk-install]: 

```
pacman -S mingw-w64-x86_64-gtk4
```

--- 

It's all, folks! 🙂
I hope I have helped with something.

Any suggestion or thoughts of how can I made it simpler, you can comment bellow.

Also, thanks @PhilippMDoerner for your support! 🎉 

[discord-thread]: https://discord.com/channels/371759389889003530/979844494205812806/1094348710521090058
[cairo-line]: https://github.com/can-lehmann/owlkettle/blob/377b7b45d3c1a234daad6f30b71caffd299c0be8/owlkettle/cairo.nim#L80
[gtk-line]: https://github.com/can-lehmann/owlkettle/blob/377b7b45d3c1a234daad6f30b71caffd299c0be8/owlkettle/gtk.nim#L27
[discord-thread-2]: https://discord.com/channels/371759389889003530/979844494205812806/1094636771368783972
[gtk-install]: https://www.gtk.org/docs/installations/windows/

**Comment by PhilippMDoerner** [comment]
I think this would be a fantastic thing to add as a page in the [nimibook docs](https://can-lehmann.github.io/owlkettle/docs/recommended_tools.html) of owlkettle.
We could change the page "Tooling" into a section "Tooling".
The current "Tooling" page could then be added to that section under the name  "GTK Tools".
The contents of what you just figured out lastly could be also a page there called "Windows MSys2" or something like that.

It would let others benefit from the time you had to invest to find these answers, which I think would be quite worthwhile.
Though I'm of course not the maintainer of the project so my word holds no sway here, just wanted to voice my opinion.

**Comment by RickBarretto** [comment]
I hope my discussion will be considered outdated soon, since I made this PR:
- #54 

Furthermore, add instructions for Windows would be awesome. And would be happy to do a PR for it too. 🙂

**Comment by can-lehmann** [comment]
This is awesome! Thank you for figuring this out and creating the PR for it :smile:

@PhilippMDoerner I also agree that this should be documented in the owlkettle docs. I think it might be worth it to create a new `Installation` section in the docs that describes the installation on Linux, Windows and MacOS (I believe some people got it working there too).

**Reply by PhilippMDoerner** [reply]
Ohhh yeah that's a way better idea. I was thinking about it in terms of "You use this with owlkettle", but from an actual user perspective this is all just setup. That makes more sense!

**Comment by RickBarretto** [comment]
BTW, a totally aside message:

@can-lehmann, Thank you for owlkettle. This project represents what a good library should be for Nim!
The use of macros and templates for make the developer's life easier, what it uses this very well.

Other projects has their own languages just to implement UI and make it declarative and easier to understand (QT with QML and Kivy with KV, for instance). But this project and Nim's metaprogramming allow us to make it seamless, with a beautiful and understandable code... 🙂

It demonstrates how metaprogramming can be used to benefit all of us.
